src/wallet/scriptpubkeyman.cpp

```diff
@@ -1717,42 +1717,160 @@ std::unordered_set<CScript, SaltedSipHasher> LegacyDataSPKM::GetScriptPubKeys()
+            CScript ms_wsh = GetScriptForDestination(WitnessV0ScriptHash(script));
+            if (HaveCScript(CScriptID(ms_wsh))) {
+                if (all_keys_compressed(keys)) {
+                    spks.insert(ms_wsh);
+                    spks.insert(GetScriptForDestination(ScriptHash(ms_wsh)));
                 }
             }
+            break;
+        }
+        }
+    }
+    // Iterate again for all the P2WSH scripts
+    for (const auto& script_pair : mapScripts) {
+        const CScript& script = script_pair.second;
+        std::vector<std::vector<unsigned char>> sols;
+        TxoutType type = Solver(script, sols);
+        if (type == TxoutType::WITNESS_V0_SCRIPTHASH) {
+            uint160 hash{RIPEMD160(sols[0])};
+            CScript witness_script;
+            int wit_ver = -1;
+            std::vector<unsigned char> wit_prog;
+            if (GetCScript(CScriptID(hash), witness_script) &&
+                !witness_script.IsPayToScriptHash() &&
+                !witness_script.IsWitnessProgram(wit_ver, wit_prog) &&
+                spks.count(witness_script) > 0) {
+                spks.insert(script);
+                spks.insert(GetScriptForDestination(ScriptHash(script)));
+            }
         }
     }
 
+    enum class ScriptContext {
+        TOP,
+        P2SH,
+        P2WSH,
+    };
+    std::function<bool(const CScript&, const ScriptContext)> is_valid_script = [&](const CScript& script, const ScriptContext ctx) -> bool {
+        std::vector<valtype> sols;
+        TxoutType spk_type = Solver(script, sols);
+
+        CKeyID keyID;
+        switch (spk_type) {
+        case TxoutType::NONSTANDARD:
+        case TxoutType::NULL_DATA:
+        case TxoutType::WITNESS_UNKNOWN:
+        case TxoutType::WITNESS_V1_TAPROOT:
+        case TxoutType::ANCHOR:
+            return ctx == ScriptContext::TOP;
+        case TxoutType::PUBKEY:
+            if (ctx == ScriptContext::P2WSH && sols[0].size() != 33) return false;
+            return true;
+        case TxoutType::WITNESS_V0_KEYHASH:
+            return ctx != ScriptContext::P2WSH;
+        case TxoutType::PUBKEYHASH:
+            if (ctx == ScriptContext::P2WSH) {
+                CPubKey pubkey;
+                if (GetPubKey(CKeyID(uint160(sols[0])), pubkey) && !pubkey.IsCompressed()) {
+                    return false;
+                }
+            }
+            return true;
+        case TxoutType::SCRIPTHASH:
+        {
+            if (ctx != ScriptContext::TOP) return false;
+            CScriptID script_id = CScriptID(uint160(sols[0]));
+            CScript subscript;
+            if (GetCScript(script_id, subscript)) {
+                return is_valid_script(subscript, ScriptContext::P2SH);
+            }
+            return true;
+        }
+        case TxoutType::WITNESS_V0_SCRIPTHASH:
+        {
+            if (ctx == ScriptContext::P2WSH) return false;
+            CScriptID script_id{RIPEMD160(sols[0])};
```

Comment:
Wouldn’t all of this be way easier and more readable, if we had explicit types for our various output script components instead of treating them as if they were just numbers and strings?

Reply:
It is possible that the solver could be rewritten to pull out all of the components of scripts into structs with proper typing and member names, but that is a much larger refactor that is out of scope for this PR.